ibc/src/client.rs

```diff
@@ -246,3 +256,90 @@ impl From<ClientConnections> for pb::ClientConnections {
         }
     }
 }
+
+// Check that the trust threshold is:
+//
+// a) non-zero
+// b) greater or equal to 1/3
+// c) strictly less than 1
+fn validate_trust_threshold(trust_threshold: TrustThreshold) -> Result<(), anyhow::Error> {
+    if trust_threshold.denominator() == 0 {
+        return Err(anyhow::anyhow!(
+            "trust threshold denominator cannot be zero"
+        ));
+    }
+
+    if trust_threshold.numerator() * 3 < trust_threshold.denominator() {
+        return Err(anyhow::anyhow!("trust threshold must be greater than 1/3"));
+    }
+
+    if trust_threshold.numerator() >= trust_threshold.denominator() {
+        return Err(anyhow::anyhow!(
+            "trust threshold must be strictly less than 1"
+        ));
+    }
+
+    Ok(())
+}
+
+// validate the parameters of an AnyClientState, verifying that it is a valid Penumbra client
+// state.
+pub fn validate_penumbra_client_state(
+    client_state: AnyClientState,
+    chain_id: &str,
+    current_height: u64,
+) -> Result<(), anyhow::Error> {
+    let tm_client_state = downcast!(client_state => AnyClientState::Tendermint)
+        .ok_or_else(|| anyhow::anyhow!("invalid client state: not a Tendermint client state"))?;
+
+    if tm_client_state.frozen_height.is_some() {
+        return Err(anyhow::anyhow!("invalid client state: frozen"));
+    }
+
+    // NOTE: Chain ID validation is actually not standardized yet. see
+    // https://github.com/informalsystems/ibc-rs/pull/304#discussion_r503917283
+    let chain_id = ChainId::from_string(&chain_id);
+    if chain_id != tm_client_state.chain_id {
+        return Err(anyhow::anyhow!(
+            "invalid client state: chain id does not match"
+        ));
+    }
+
+    // check that the revision number is the same as our chain ID's version
+    if tm_client_state.latest_height.revision_number != chain_id.version() {
+        return Err(anyhow::anyhow!(
+            "invalid client state: revision number does not match"
+        ));
+    }
+
+    // check that the latest height isn't gte the current block height
+    if tm_client_state.latest_height.revision_height >= current_height {
+        return Err(anyhow::anyhow!(
+                "invalid client state: latest height is greater than or equal to the current block height"
+            ));
+    }
+
+    // check client proof specs match penumbra proof specs
+    let penumbra_proof_specs: ProofSpecs = vec![jmt::ics23_spec()].into();
+    if penumbra_proof_specs != tm_client_state.proof_specs {
+        return Err(anyhow::anyhow!(
+            "invalid client state: proof specs do not match"
+        ));
+    }
+
+    // check that the trust level is correct
+    validate_trust_threshold(tm_client_state.trust_level)?;
+
+    // TODO: check that the unbonding period is correct
+    //
+    // - check unbonding period is greater than trusting period
+    if tm_client_state.unbonding_period < tm_client_state.trusting_period {
```

Comment:
What's the relation between `tm_client_state.unbonding_period` and the `unbonding_epochs` in our chain parameters? We need these to be aligned, right?